schema_compatibility.go

```diff
@@ -230,12 +232,192 @@ func (c *SchemaCompatibility) contains(a []string, s string) bool {
 	return false
 }
 
-func (c *SchemaCompatibility) getField(a []*Field, f *Field) (*Field, bool) {
+type getFieldOptions struct {
+	fieldAlias bool
+	elemAlias  bool
+}
+
+func (c *SchemaCompatibility) getField(a []*Field, f *Field, optFns ...func(*getFieldOptions)) (*Field, bool) {
+	opt := getFieldOptions{}
+	for _, fn := range optFns {
+		if fn == nil {
+			continue
+		}
+		fn(&opt)
+	}
 	for _, field := range a {
 		if field.Name() == f.Name() {
 			return field, true
 		}
+		if opt.fieldAlias {
+			for _, alias := range f.Aliases() {
+				if field.Name() == alias {
+					return field, true
+				}
+			}
+		}
+		if opt.elemAlias {
+			for _, alias := range field.Aliases() {
+				if f.Name() == alias {
+					return field, true
+				}
+			}
+		}
 	}
 
 	return nil, false
 }
+
+func isNative(typ Type) bool {
+	switch typ {
+	case Null, Boolean, Int, Long, Float, Double, Bytes, String:
+		return true
+	default:
+	}
+
+	return false
+}
+
+func isPromotable(typ Type) bool {
+	switch typ {
+	case Int, Long, Float, String, Bytes:
+		return true
+	default:
+	}
+
+	return false
+}
+
+func (c *SchemaCompatibility) Resolve(reader, writer Schema) (Schema, error) {
+	if reader.Type() == Ref {
+		reader = reader.(*RefSchema).Schema()
+	}
+	if writer.Type() == Ref {
+		writer = writer.(*RefSchema).Schema()
+	}
+
+	if err := c.compatible(reader, writer); err != nil {
+		return nil, err
+	}
+
+	if writer.Type() != reader.Type() {
+		if isPromotable(writer.Type()) {
+			return reader, nil
```

Comment:
This is not correct. When there is a promotion, the data is encoded in the former and typed by the latter. The way it stands here, the codec would try to decode the readers type, but it was written in the writers type. This basically breaks the decoder. 

It is the single biggest issue I encountered when thinking about this topic.

Reply:
So that's the blocking points ^^' Thanks for pointing it out! I should start thinking about it first.

This Java implementation might be a source of inspiration [code](https://github.com/apache/avro/blob/main/lang/java/avro/src/main/java/org/apache/avro/io/ResolvingDecoder.java)